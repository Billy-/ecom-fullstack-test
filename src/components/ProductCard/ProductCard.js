import React from 'react';
import { string, number, shape, func } from 'prop-types';
import { withRouter } from 'react-router';

import Card from '../Card';

import styles from './ProductCard.scss';

const ProductCard = ({
    title,
    image,
    productLabel,
    description,
    priceLabel,
    currency,
    price,
    ctaLink,
    cta,
    router
}) => (
    <Card className={styles.container}>
        <img
            className={styles.image}
            src={image.path}
            alt={image.alt}
        />
        <div className={styles.content}>
            {productLabel && (
                <span className={styles.productLabel}>{productLabel}</span>
            )}
            <h3 className={styles.title}>{title}</h3>
            <p className={styles.description}>{description}</p>
        </div>
        <div className={styles.footer}>
            {/* TODO: Extract Price into /components */}
            <span className={styles.priceContainer}>
                <span className={styles.priceLabel}>{priceLabel}</span>
                <span className={styles.price}>{currency}{(price / 100).toFixed(2)}</span>
            </span>
            {/* TODO: Extract CTA into /components */}
            <div
                className={styles.cta}
                onClick={() => router.push(ctaLink)}
            >{cta}</div>
        </div>
    </Card>
);

ProductCard.propTypes = {
    title: string.isRequired,
    image: shape({
        path: string.isRequired,
        alt: string.isRequired
    }).isRequired,
    productLabel: string.isRequired,
    description: string.isRequired,
    priceLabel: string.isRequired,
    currency: string.isRequired,
    price: number.isRequired,
    ctaLink: string.isRequired,
    cta: string.isRequired,
    router: shape({
        push: func.isRequired
    }).isRequired
};

export default withRouter(ProductCard);
