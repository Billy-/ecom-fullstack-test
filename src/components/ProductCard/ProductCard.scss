@import '../../sass/variables';

$padding: 20px;

.container {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding-bottom: $padding;
}

.image {
  max-width: 100%;
  margin: 0 auto;
  display: block;
}

.productLabel {
  background: $red;
  color: $white;
  padding: 5px 10px;
  padding-left: $padding;
  font-weight: bold;
  // The reason I made this position absolute is
  // so that each card will still be the same height,
  // regardless of whether this element is present or not.
  position: absolute;
  top: 0;
  left: 0;
}

.content {
  position: relative;
  padding: 0 $padding;
  padding-top: 30px; // For abs pos productLabel
}

.title {
  margin-bottom: 0;
  font-size: 24px;
}

.description {
  margin: ($padding * 2) 0;
  font-size: 20px;
  line-height: 24px;
}

.footer {
  padding: 0 $padding;
}

.priceContainer {
  display: block;
  margin-bottom: $padding;
}

.priceLabel {
  color: $grey;
  font-size: 14px;
  margin-right: 10px;
}

.price {
  font-weight: bold;
  font-size: 24px;
}

.cta {
  border: 6px solid $green;
  font-weight: bold;
  padding: 10px 20px;
  text-align: center;
  cursor: pointer;
}
