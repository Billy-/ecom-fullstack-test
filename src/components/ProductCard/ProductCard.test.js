import React from 'react';
import renderer from 'react-test-renderer';

import ProductCard from './ProductCard';

const props = {
    title: 'foo',
    image: {
        path: 'foo',
        alt: 'foo'
    },
    productLabel: 'foo',
    description: 'foo',
    priceLabel: 'foo',
    currency: '$',
    price: 9999,
    ctaLink: 'foo',
    cta: 'foo',
    history: {
        push: () => {}
    }
};

describe('ProductCard', () => {
    it('renders without throwing', () => {
        renderer.create(<ProductCard {...props} />);
    });

    it('renders an image with correct path and alt props', () => {
        const instance = renderer.create(<ProductCard {...props} />).root;
        const imgProps = instance.findByType('img').props;
        expect(imgProps.src).toBe(props.image.path);
        expect(imgProps.alt).toBe(props.image.alt);
    });

    it('puts the currency symbol before the price', () => {
        const instance = renderer.create(<ProductCard {...props} />).root;
        const renderedCurency = instance.findByProps({ className: 'price' }).props.children[0];
        expect(renderedCurency).toBe(props.currency);
    });

    it('puts a decimal in the right place in the price', () => {
        const instance = renderer.create(<ProductCard {...props} />).root;
        const renderedPrice = instance.findByProps({ className: 'price' }).props.children[1];
        expect(renderedPrice).toBe('99.99');
    });

    // TODO: Finish writing tests
    // it('renders XYZ')

    it('matches snapshot', () => {
        const tree = renderer
          .create(<ProductCard {...props} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
