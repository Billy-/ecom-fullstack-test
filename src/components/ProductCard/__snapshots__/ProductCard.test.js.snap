// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard matches snapshot 1`] = `
<div
  className="card container"
>
  <img
    alt="foo"
    className="image"
    src="foo"
  />
  <div
    className="content"
  >
    <span
      className="productLabel"
    >
      foo
    </span>
    <h3
      className="title"
    >
      foo
    </h3>
    <p
      className="description"
    >
      foo
    </p>
  </div>
  <div
    className="footer"
  >
    <span
      className="priceContainer"
    >
      <span
        className="priceLabel"
      >
        foo
      </span>
      <span
        className="price"
      >
        $
        99.99
      </span>
    </span>
    <div
      className="cta"
      onClick={[Function]}
    >
      foo
    </div>
  </div>
</div>
`;
